src/masonite/notification/drivers/SlackDriver.py

```diff
@@ -0,0 +1,117 @@
+"""Slack notification driver"""
+import requests
+
+from ...exceptions import NotificationException
+from .BaseDriver import BaseDriver
+
+
+class SlackDriver(BaseDriver):
+
+    WEBHOOK_MODE = 1
+    API_MODE = 2
+    send_url = "https://slack.com/api/chat.postMessage"
+    channel_url = "https://slack.com/api/conversations.list"
+
+    def __init__(self, application):
+        self.application = application
+        self.options = {}
+        self.mode = self.WEBHOOK_MODE
+
+    def set_options(self, options):
+        self.options = options
+        return self
+
+    def send(self, notifiable, notification):
+        """Used to send the notification to slack."""
+        slack_message = self.build(notifiable, notification)
+        if slack_message._mode == self.WEBHOOK_MODE:
+            self.send_via_webhook(slack_message)
+        else:
+            self.send_via_api(slack_message)
+
+    # def queue(self, notifiable, notification):
+    # TODO
+    #     """Used to queue the notification to be sent to slack."""
+    #     method, payload = self.prepare(notifiable, notification)
+    #     return self.application.make("queue").push(method, args=payload)
+
+    def build(self, notifiable, notification):
+        """Build Slack message payload sent to Slack API or through Slack webhook."""
+        slack_message = self.get_data("slack", notifiable, notification)
+        recipients = self.get_recipients(notifiable)
+        mode = self.get_sending_mode(recipients)
+        slack_message = slack_message.mode(mode)
+
+        if mode == self.WEBHOOK_MODE:  # and not slack_message._webhook:
+            slack_message = slack_message.to(recipients)
+        elif mode == self.API_MODE:
+            # if not slack_message._channel:
+            slack_message = slack_message.to(recipients)
+            if not slack_message._token:
+                slack_message = slack_message.token(self.options.get("token"))
+        return slack_message
+
+    def get_recipients(self, notifiable):
+        recipients = notifiable.route_notification_for("slack")
+        if not isinstance(recipients, (list, tuple)):
+            recipients = [recipients]
+        return recipients
+
+    def get_sending_mode(self, recipients):
+        modes = []
+        for recipient in recipients:
+            if recipient.startswith("https://hooks.slack.com"):
+                modes.append(self.WEBHOOK_MODE)
+            else:
+                modes.append(self.API_MODE)
+        if len(set(modes)) > 1:
+            raise NotificationException("Slack sending mode cannot be mixed.")
+        return modes[0]
+
+    def send_via_webhook(self, slack_message):
```

Comment:
Whats the benefit of having both sending via webhook and sending via API?

Reply:
Based on our Slack discussions. Both modes are useful (have pros and cons). Maybe later we could deprecate one or another ...